packages/frontpage/lib/data/db/notification.ts

```diff
@@ -0,0 +1,140 @@
+import "server-only";
+
+import { cache } from "react";
+import { db } from "@/lib/db";
+import * as schema from "@/lib/schema";
+import { and, eq, lt, desc, isNull, count } from "drizzle-orm";
+import { invariant } from "@/lib/utils";
+import { ensureUser } from "../user";
+import { DID } from "../atproto/did";
+
+declare const tag: unique symbol;
+export type Cursor = { readonly [tag]: "Cursor" };
+
+function cursorToDate(cursor: Cursor): Date {
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  return new Date(cursor as any);
+}
+
+function createCursor(date: Date): Cursor {
+  return date.toISOString() as unknown as Cursor;
+}
+
+export type Notification = Awaited<
+  ReturnType<typeof getNotifications>
+>["notifications"][number];
+
+export const getNotifications = cache(
+  async (limit: number, cursor: Cursor | null) => {
+    const user = await ensureUser();
+
+    const joins = await db
+      .select()
+      .from(schema.Notification)
+      .where(
+        and(
+          eq(schema.Notification.did, user.did),
+          cursor
+            ? lt(schema.Notification.createdAt, cursorToDate(cursor))
+            : undefined,
+        ),
+      )
+      .leftJoin(
```

Comment:
comments and posts should be inner joins to prevent null info

Reply:
I don't think so, I want left join here so that I can have either a post or a comment row attached to the notification. Inner join would require both a post and a comment to exist for this notification I think which is impossible.

Reply:
a comment always has a post and a post always has a comment in our current iteration of notifications